### ISSUE TITLE
Dropbox refresh token

### ISSUE THREAD

**Issue by alicetraverso27** [node-red/node-red-web-nodes#304]
Hi everyone! 
I'm having trouble with the latest update for Dropbox node.
The redirect link assigned to 4th step of the how-to procedure has http protocol, BUT according to Dropbox guide lines it has to be https, otherwhise it won't be accepted in the "Redirect URIs" section of the app.

![Schermata 2022-11-25 alle 11 54 03](https://user-images.githubusercontent.com/62282840/203971182-26134dca-c80c-4d8c-b9dd-0b42416c06ef.jpg)

![Schermata 2022-11-25 alle 12 03 55](https://user-images.githubusercontent.com/62282840/203971885-74ab0857-c5ac-4e83-ab8d-f0988095479c.png)

So, basically, I keep getting a 400 error response when trying to create my refresh token.

![Schermata 2022-11-25 alle 11 49 19](https://user-images.githubusercontent.com/62282840/203971956-2c8cccdc-39da-4707-82b7-b7b496d9e984.png)

Any suggestions on how to solve this? 

Thank you in advance!

**Comment by Paul-Reed** [comment]
I seem to recall that if you do not use https, you can use a `Localhost` path to substitute the `redirect_uri`.
Please let us know your outcome, and we'll update the readme.

**Comment by alicetraverso27** [comment]
My nodered installation is served on https protocol so I can't see why the redirect_uri is generetad with http instead of https. 
And sadly I have no privileges to run my flow locally so I can't use a localhost path :(

**Comment by Paul-Reed** [comment]
Can you tell us in more detail about your system please.
How & where are you running node-RED, operating system, etc...

**Comment by alicetraverso27** [comment]
node-Red v.3.0.2
cloudflare https —> nginx https self signed proxy --> nodered:latest in docker

**Comment by bartbutenaers** [comment]
Hi @alicetraverso27,
Perhaps there is a bug in the [endpoint](https://github.com/node-red/node-red-web-nodes/blob/master/dropbox/dropbox.js#L335) that generates the redirect url.  Not sure at the moment...

A few questions:
1. Based on your feedback, I assume there is a "http***s***" in your adresbar when you see your flow editor.  Is that correct? 
2. Have you tried to enter the redirect url in the dropbox website, but with http***s*** instead of http?

Bart

**Comment by dceejay** [comment]
@bartbutenaers 
totally guessing but does nginx change the https request "outside" to plain http on the inside - which is what the nodee then sees ?

**Comment by bartbutenaers** [comment]
@dceejay,
yes indeed.  That is the same thing I was thinking when I saw Nginx being mentioned.  But wanted to make sure that is the case.

If this is the case, then I am not sure how to solve this is in a user-friendly way.  Of course I could say: if not localhost is being used, then I always use https (because that is what Dropbox requires).  But of course when people have not setup ssl for Node-RED, then it won't work either because Dropbox cannot reach their Node-RED system via that https redirect url.  

We could of course add a red description:

![image](https://user-images.githubusercontent.com/14224149/204092884-4bdbca4a-3ead-43b9-8f4e-073d472cd5b7.png)

But of course it all depends on whether this is the issue for @alicetraverso27 ...

P.S. Found this quote somewhere, which is rather useful information: *"OAuth 2 redirect URIs don't technically need to be publicly accessible. They just need to be accessible to the user going through the OAuth 2 app authorization flow. "*

**Comment by stephennutt** [comment]
Was there any further investigation into this because I am having similar problem:

If I click the link on the Node Setup page, I get this:
![image](https://user-images.githubusercontent.com/25011547/209886794-f291eba8-2e8d-4fb3-9e15-6720982f746f.png)

If I instead go to localhost:1880/dropbox/authenticate directly in a browser, I get this:
![image](https://user-images.githubusercontent.com/25011547/209886829-5ea1bdfb-76d8-45b9-b617-4c3263f7e6b6.png)

I believe I have it setup correctly in Dropbox:
![image](https://user-images.githubusercontent.com/25011547/209886880-0058ea27-e6d5-46e9-906a-3043214396b6.png)

Any suggestions?  If not, is there a way to revert to previous version of the pallet?

**Comment by bartbutenaers** [comment]
HI @stephennutt,

No the original question was not solved, since the OP didn't answer any of my feedback questions...

You get a entirely other error message, so it would have been better if you had created a new issue here on Github...
I have a complete lack of free time, so I can only "try" to solve it if you can help me a bit with it.
However if you want to revert to the old version (like some others already did), fine for me but then it won't be solved.

Your error is thrown [here](https://github.com/node-red/node-red-web-nodes/blob/master/dropbox/dropbox.js#L398) in the code.  So if you could do this:
1. Find the dropbox.js file on your raspberry (somewhere in ~.node-red/node_modules/...)
2. Just before the failing line (398), add this code snippet:
   ```
   // Perhaps the 'req' will show too much output, and is it enough to use 'req.query'.  Don't know...
   let str = JSON.stringify(req, null, 4); // indented output.
   console.log(str); // Logs output to dev tools console.
   ```
3. Restart Node-RED
4. Repeat your actions for dropbox
5. Share with us the extra stuff in the console log of your Node-RED.

But even if we can fix this, we still need to find the root cause why you get that error.  Because the code that fails is at the end of your authentication flow, i.e. after you got the ***invalid_client_id***.  Can you explain why you use localhost, while your screenshot shows raspberrypi.beta?  

Perhaps @Paul-Reed can provide some useful information here...

Bart

**Comment by Paul-Reed** [comment]
@bartbutenaers - Stephen appears to be using the URI 'http://raspberrypi.beta:1880/dropbox/authenticate' in the node config (because that URI is auto generated), whilst the dropbox 'app' is using 'http://localhost:1880/dropbox/authenticate', which should really result in the same result, but...
I seem to recall that localhost worked for you during early testing, when the same localhost URI was specified in both places. To test that, is there a way that Stephen could hardcode the localhost URI in the node, to test, and see if that makes the difference.
Somewhere around https://github.com/node-red/node-red-web-nodes/blob/master/dropbox/dropbox.html#L68-L79 ?

Is that worth trying?

**Comment by stephennutt** [comment]
Sorry, I don't know much about how this node works behind the scenes so I assumed that my issue was related to OP's.

The issue is that the edit node suggests using http://192.168.1.208:1880/dropbox/authenticate but Dropbox will not let you enter a http URL unless it is localhost.  192.168.1.208 is the device that is hosting Node Red and I also have a DNS to it of raspberrypi.beta.  That's why that shows up on previous post. 

![image](https://user-images.githubusercontent.com/25011547/209946579-8c089ed0-c8c4-4288-9bc3-1e712677e829.png)
 

I added 3 variations of the redirect URLs in Dropbox:
![image](https://user-images.githubusercontent.com/25011547/209946793-b6b8ee99-5cd2-4761-99a6-ef74e090cf38.png)


 Now, if I click on the authentication flow link in the Node setup, I get this:
![image](https://user-images.githubusercontent.com/25011547/209946870-2979c745-5575-40c9-a372-ec4117f5fa6a.png)

If I enter either of the https URLs in a web browser using the Raspberry Pi, I get "Secure Connection Failed".  If I open the http localhost URL, I get the Syntax error I shared previously.  Sorry, I don't know enough about the inner workings of Node Red to be more help.

**Comment by Paul-Reed** [comment]
@stephennutt , have you carried out the actions suggested by @bartbutenaers in his last post yet??
Please post the response.

**Comment by stephennutt** [comment]
I looked for dropbox.js in the /home/pi/.node-red/node_modules/dropbox folder and found 3:

In the folder:     /src -- only 211 lines
In the folder:     /es/src folder: -- only 235 lines
in the folder:     /cjs/src -- only 247 lines

None of them had the 398+ lines that he mentioned.

**Comment by Paul-Reed** [comment]
Try `~/.node-red/node_modules/node-red-node-web-nodes/dropbox/dropbox.js`

**Comment by stephennutt** [comment]
Is there an easy way to revert to previous version of the pallet?  In my inexperienced investigation of this issue and trouble-shooting, I almost lost my entire Node Red installation and I am not willing to risk screwing that up.

Thank you for your help.  I am sure it is a PEBCAK problem.

**Comment by bartbutenaers** [comment]
Ok, thanks @Paul-Reed and @stephennutt for the extra details.
Did quickly a bit of homework to refresh my brain.

Not very structured yet, but we need to start somewhere...

1. When @stephennutt calls the redirect_url (`localhost:1880/dropbox/authenticate`) manually, then our endpoint fails.  That is ***normal*** behaviour.  Because in a normal authentication flow, the Dropbox authentication app sends a `req.query` object to that endpoint, containing all kind of information about the result of the authentication flow.  When you call this url manually, there is no such data in the request, so the endpoint cannot parse it and will fail...

2. What I don't get is that @stephennutt gets an `invalid client_id` (in his first report above) followed by his redirect_url:

   ![image](https://user-images.githubusercontent.com/14224149/210012645-c4fdb939-d2a2-4de2-8a6e-b258a6c734d2.png)

   The _**client id**_ is in fact the _**app id**_ that you need to copy manually from your dropbox account website to the config screen of the dropbox node.  I had expected to see an incorrect app id, instead of the redirect url.  Something like this:

   ![image](https://user-images.githubusercontent.com/14224149/210012781-ab7fde1e-6ac2-4d93-9740-c30549af9dd3.png)

   Or the below error, when you have forgotten to copy manually your app id in the config screen of this node:

   ![image](https://user-images.githubusercontent.com/14224149/210012847-05f77c61-e92f-488d-94c0-38ca03112f37.png)

   I assume that Stephen hasn't accidentally copied his redirect_url in the app_key input field of his dropbox node config screen...  So not sure then why the client_id and redirect_uri are mixed up here.  Don't see anything in this node at first sight that could go wrong.

   BTW we should list all these 'known' errors on our readme page.

3. We discussed last week on Discourse whether it would be possible to setup SSL with a self-signed certificate.  However from this [discussion](https://www.dropboxforum.com/t5/Dropbox-API-Support-Feedback/Test-webhooks-implementation-with-self-signed-certificate/m-p/7588/highlight/true#M338) it becomes (obvious) clear why the Dropbox SDK does ***not*** allow this:
   ```
   We do certificate validation to ensure we're talking to the right server. (Otherwise a man-in-the-middle attack could intercept your webhooks.) So for now you'll need to use a non-self-signed cert for that domain.
   But we'll track this as a feature request.
   ```
   Not sure whether this feature has been implemented.  If so, we could allow the certification validation to be disabled.  This will be less secure, but I assume that is only applicable if you do all this in a flow editor that is opened from a computer outside your LAN.  And that you call the (public) redirect_url over https.  Which would be rather stupid...  When you do this kind of stuff on your LAN, I assume that skipping certification validation is not a big security concern?

   Anyway at this moment, you will need to have a certificate issued by a certified authority (e.g. LetsEncrypt).  Which makes thing yet more complex for our users...

4. From the Dropbox OAuth [documentation](https://developers.dropbox.com/oauth-guide), I see that calling a redirect_url is optional:
   > The redirect_uri is optional with the code flow - if unspecified, the authorization code is displayed on dropbox.com for the user to copy and paste to your app.  This extra step is less convenient for end users, but appropriate for apps that cannot support a redirect.

   When we would skip this step, that would result in an extra manual step for our users.   Because they need to get the code from dropbox.com instead of from our popup window.   ***But perhaps I should make the redirect uri optional, for people that don't have SSL (with a certificate signed by a CA).***  Then at least it works, but of course it is less user friendly of course...

5. Stephen has now added multiple redirect uri's to his dropbox account.  Dropbox checks that the specified redirect_uri parameter in our Authorization URL matches one of those registered values.  However I [pass](https://github.com/node-red/node-red-web-nodes/blob/master/dropbox/dropbox.html#L90) the original authentication url (which is auto generated and displayed in the config screen) to the popup window.  In Stephen's screenshot that original url is http://192.168.1.208, which is ***not*** in the list of allowed redirect uri's in his dropbox account (because http is not allowed for non-localhost).  So it is pretty normal that he sees the error *"it must exactly match one of the redirect URI's..."*.  I don't think it is useful that I make this auto generated redirect uri manually adjustable by the user (by displaying it into a input text field)?  But I might be mistaken.  My head has become blurry...

My time is up for today.  Lot's of thing to think about.
All idea's are welcome!

**Comment by bartbutenaers** [comment]
Morning guys,
I had a look at [this](https://www.dropboxforum.com/t5/Dropbox-API-Support-Feedback/python3-and-API-V2-turn-off-certificate-checking/td-p/408191) dropbox discussion, where somebody has a feature request, to be able to skip certificate validation in the Dropbox Python SDK.  Of course we use their Javascript SDK, but the principle is the same.

Summarized: Dropbox is not going to implement this, dus to security concerns.  Which I can understand, for cases where you access your (public) redirect url from outside your LAN.  Their proposal is to add the certificate to the list of trusted certificates on your system.  But of course asking our users to do this, would be much more complex compared to just get the new generated refresh token from your account webpage on dropbox.com (instead from our last screen)...

So therefore ***I suggest to drop the whole redirect_uri mechanism***:

![image](https://user-images.githubusercontent.com/14224149/210042044-69145904-3505-4cd8-8dbf-628303d41178.png)

Which would include following changes:
+ Drop the endpoint A that currently generates a redirect_uri
+ Don't call that endpoint A anymore when the node's config screen is opened, so don't show it anymore (as step 4) in the config screen.
+ Don't pass the redirect anymore to the Dropbox Javascript SDK.
+ Change the text on the config screen (step 6) to tell the users that their generated refresh token need to be fetched from their account on dropbox.com (instead of copying it from the last screen C.
+ Remove endpoint C that generates the html for screen C (where I displayed the refreshed token), because the authentication flow won't call it anymore.

Another advantage is that the flow becomes much easier to follow/troubleshoot and the codebase of this node is easier to maintain.

Of course I haven't tested any of this yet!  And my updated texts in the screenshot are not final, just to demonstrate what I mean.

Then I think it will work for most users.  I didn't have these problems during development, because I use LetsEncrypt certificates.  But it seems - even after the attack of hackers on lots of Node-RED installations - that people keep on working without security.  We simply will have to learn to live with that...

I will do the changes, since we have no other option I am afraid.  
Would be nice if anybody could help me with the test work, since I don't use Dropbox myself...

**Comment by bartbutenaers** [comment]
I think I have a solution.  But need to have some input about it should look like on the screen.
So I have created a public discussion about it: https://discourse.nodered.org/t/refresh-tokens-in-node-red-node-dropbox-now-published-as-v2-0-0/65610

**Comment by Paul-Reed** [comment]
Here's the discussion. 

https://discourse.nodered.org/t/input-needed-for-a-fix-in-the-dropbox-node-refresh-tokens/72949
